### ISSUE TITLE
RSpec clear_examples almost works in a complex setup with Runner.run in a loop.

### ISSUE THREAD

**Issue by guyboertje** [rspec/rspec-core#2367]
In the [Logstash](https://github.com/elastic/logstash) product we use rake and rspec to run many specs on the main project and the 100 or so plugins that have their own rspec suite.

We thank the rspec community for their continued work on this important library.

When run individually each plugin spec suite runs fine with the set of helpers required correctly.

If I attempt to automate these with a loop using `RSpec::Core::Runner.run(rspec_args)` and `RSpec.clear_examples` it almost works.

1) the deprecation warnings are not cleared and continue to be reported in a cumulative manner (i know but we have tech debt and a smallish team). I was able to use `--deprecation-out` to reduce the output in the console.
2) the shared_examples from `require ..../shared_examples` are removed by the `world.reset` method doing `@shared_example_group_registry = nil` NOTE: commenting this line out worked for me.
3) the `--out` setting did not work because the IO is closed after the first loop iteration and `clear_examples` does not reopen the IO. I had to pipe the stdout output to a file.

It is important in our case to have the output go to a file because with JRuby stack traces etc. we can't always scroll back in the console.

It is debatable whether the randomized seed should be reset by `clear_examples`.

I tried to change some code with an eye to a PR (its always nice to get fixes with issues, we know) but the finding of how to clear the calls in the `DeprecationReporterBuffer` defeated me. 

May I suggest a method called something like `reset_before_rerun` that is `clear_examples` and more? It would better reveal the intention of the functionality.

Previous discussion - https://github.com/rspec/rspec-core/issues/1700

**Comment by xaviershay** [comment]
nice find, thanks for detailed report!

**Comment by xaviershay** [comment]
Could you please link up an example of your current usage? I can't seem to find it :(

Just to check, `RSpec.reset` doesn't do what you want, right? Presumably because you want to preserve configuration.

**Comment by xaviershay** [comment]
I think the attached PR addresses all your issues. Would you mind trying it out to see?

**Comment by guyboertje** [comment]
@xaviershay 
Re example: its a complex setup and hard to setup. However, this [line in our test.rake](https://github.com/elastic/logstash/blob/master/rakelib/test.rake#L63) is the starting point.
This is my modified version that loops on `Runner.run`
```rake
  desc "run all installed plugins specs in a separate rspec run, args:['format']"
  task "plugins-separate", [:format] => ["setup"] do  |t, args|
    format = args.format || "documentation"
    # additionally we could have an output file path as an arg
    exit_code = 0

    plugins_to_exclude = ENV.fetch("EXCLUDE_PLUGIN", "").split(",")
    # grab all spec files using the live plugins gem specs. this allows correclty also running the specs
    # of a local plugin dir added using the Gemfile :path option. before this, any local plugin spec would
    # not be run because they were not under the vendor/bundle/jruby/2.0/gems path

    STDERR.puts "starting..."
    i = 1
    test_files = LogStash::PluginManager.find_plugins_gem_specs.shuffle.each do |spec|

      if plugins_to_exclude.size > 0 && plugins_to_exclude.include?(Pathname.new(spec.gem_dir).basename.to_s)
        next
      end

      test_files = Rake::FileList[File.join(spec.gem_dir, "spec/{input,filter,codec,output}s/*_spec.rb")]

      rspec_args = ["--color", "--order", "rand", "--deprecation-out", "data/deprecations-out.txt", "--format", format, test_files]

      STDERR.puts "----------------- #{i} Running #{spec.gem_dir}"

      begin
        exit_code += RSpec::Core::Runner.run(rspec_args)
      rescue => e
        STDERR.puts e.message
        STDERR.puts e.backtrace.take(8)
        STDERR.puts "----------------- #{i}"
      end

      RSpec.clear_examples
      i = i.succ
    end

    STDERR.puts "done..."
    exit(exit_code == 0 ? 0 : 1)
  end
```
Re `RSpec.reset`, tried it but as you suspect configuration that is modified with `require "logstash/devutils/rspec/spec_helper"` is lost.
It would be cool to be able to have a callable be registered with the configuration (once in a require) and this callable is called for each run. We would be less at the mercy of require vs load.

**Comment by guyboertje** [comment]
I will attempt to try the PR.

**Comment by guyboertje** [comment]
@xaviershay 
I tried your PR.

1) Deprecation accumulation: fixed.
2) RSpec configuration issues: fixed IMO.
3) `--out` IO early close: fixed.

However, because the [file is opened with "w"](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb#L252) each run overwrites the previous one.
Changing "w" to "a" worked for me but I am not sure how this affects other use cases.

Further observations, when doing multiple runs like this, especially with the progress format, its not easy to correlate loop info (`spec.gem_dir` from above example) with the text in the `--out` file.
In the loop I tried to write  `"----------------- #{i} Running #{spec.gem_dir}"` to the `--out` file using:
`RSpec.configuration.output_stream` but this was set to $stdout;
`RSpec.configuration.formatters.first.output`, this works but incurs the cost of [formatters.dup](https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/configuration.rb#L896) for each call.

I considered suggesting adding a preamble command line arg that would be output with the run options prelude but I think that would get messy.

I am not sure you should change this because now that the IO is not closed The RSpec team advice should probably be to create, use and inject ones own IO instance to the run method for these complex cases, e.g. `RSpec::Core::Runner.run(rspec_args, $stderr, io)`. I tested this with File.new and `"a"` everything works as desired.

**Comment by xaviershay** [comment]
Thank you!

> Changing "w" to "a" worked for me but I am not sure how this affects other use cases.

In regular usage (running the same suite multiple times), I'd expect the file to be overwritten as it is now.

> The RSpec team advice should probably be to create, use and inject ones own IO instance to the run method for these complex cases, e.g. RSpec::Core::Runner.run(rspec_args, $stderr, io).

Yup :) Out of curiosity, how were you configuring it before? With a command line option?

**Comment by guyboertje** [comment]
@xaviershay 

You're welcome.

Ignore the "a" - I think the advice should be "create and inject your own IO instead of using --out"

and yes, before using my own IO instance I used the --out in the rspec_args array.

In terms of advice I mean in the relishapp doc page [running-specs-multiple-times-with-different-runner-options-in-the-same-process](https://www.relishapp.com/rspec/rspec-core/v/3-5/docs/running-specs-multiple-times-with-different-runner-options-in-the-same-process)

**Comment by JonRowe** [comment]
I'm going to mark this as closed in preference for #2368
